apps/site/assets/ts/schedule/components/schedule-finder/ScheduleFinderModal.tsx

```diff
@@ -0,0 +1,103 @@
+import React, { ReactElement, useState } from "react";
+import { DirectionId, Route } from "../../../__v3api";
+import {
+  SimpleStopMap,
+  RoutePatternsByDirection,
+  ServiceInSelector,
+  ScheduleNote as ScheduleNoteType,
+  SelectedOrigin
+} from "../__schedule";
+import Modal from "../../../components/Modal";
+import OriginModalContent from "./OriginModalContent";
+import ScheduleModalContent from "./ScheduleModalContent";
+
+export type Mode = "origin" | "schedule";
+
+interface Props {
+  closeModal: () => void;
+  directionChanged?: (direction: DirectionId) => void;
+  initialDirection: DirectionId;
+  initialMode: Mode;
+  initialOrigin: SelectedOrigin;
+  originChanged?: (origin: SelectedOrigin) => void;
+  route: Route;
+  routePatternsByDirection: RoutePatternsByDirection;
+  scheduleNote: ScheduleNoteType | null;
+  services: ServiceInSelector[];
+  stops: SimpleStopMap;
+  today: string;
+}
+
+export default ({
+  closeModal,
+  directionChanged,
+  initialDirection,
+  initialMode,
+  initialOrigin,
+  originChanged,
+  route,
+  routePatternsByDirection,
+  scheduleNote,
+  services,
+  stops,
+  today
+}: Props): ReactElement => {
+  const [direction, setDirection] = useState(initialDirection);
+  const [mode, setMode] = useState(initialMode);
+  const [origin, setOrigin] = useState(initialOrigin);
+
+  const handleChangeDirection = (newDirection: DirectionId): void => {
+    setDirection(newDirection);
+    setOrigin(null);
+    setMode("origin");
+    if (directionChanged) directionChanged(newDirection);
+    if (originChanged) originChanged(null);
+  };
+
+  const handleChangeOrigin = (newOrigin: SelectedOrigin): void => {
+    setOrigin(newOrigin);
+    setMode("schedule");
+    if (originChanged) originChanged(newOrigin);
+  };
+
+  return (
+    <Modal
+      focusElementId={mode === "origin" ? "origin-filter" : "modal-close"}
+      ariaLabel={{
+        label:
+          mode === "origin"
+            ? "Choose Origin Stop"
+            : `Schedules to ${route.direction_names[direction]}`
+      }}
+      className={mode === "origin" ? "schedule-finder__origin-modal" : ""}
+      closeModal={closeModal}
+    >
+      {() => (
+        <>
+          {mode === "origin" && (
+            <OriginModalContent
```

Comment:
@digitalcora I'm thinking here and below, let's extract the TSX markup into functions, so that the markup doesn't drown out the control-flow logic. I mean something like:

```
const originModalContent = () => 
  <OriginModalContent 
    handleChangeOrigin={handleChangeOrigin}
    ...
/>

const scheduleModalContent = () =>
  <ScheduleModalContent
    handleChangeDirection={handleChangeDirection}
    ...
  />

{mode === "origin" && originModalContent()}
{mode === "schedule" && origin !== null && scheduleModalContent()}
```

Reply:
As-written, this wouldn't compile because the `ScheduleModalContent` component needs to have an origin defined, but in the context of the larger `ScheduleFinderModal` the origin is allowed to be null. But beyond that, I think this could have some unwanted side effects because we'd be "rendering" both of the components no matter what the modal mode is, instead of just the one that corresponds to the current mode (and having the other one not exist at all).

Reply:
@digitalcora right, that's why they're anonymous functions which return React elements, rather than React elements themselves--we only call the one, if any, that's appropriate.

Reply:
I see, but that doesn't resolve the issue of required values possibly not existing.

Reply:
true, in this case I think it would be OK to use the `!` modifier for the possibly-null values within the function bodies to force the issue, as we're assuring that at the callsite.

Reply:
I've resolved this per our in-person conversation, hopefully it's more readable now!